jetty-websocket/javax-websocket-server-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/server/deploy/WebSocketServerContainerInitializer.java

```diff
@@ -143,7 +143,7 @@ public static ServerContainer configureContext(ServletContextHandler context) th
         
         // Build HttpClient
         HttpClient httpClient = (HttpClient) context.getServletContext().getAttribute(HTTPCLIENT_ATTRIBUTE);
-        if(httpClient == null)
+        if ((httpClient == null) && (context.getServer() != null))
         {
             httpClient = (HttpClient) context.getServer().getAttribute(HTTPCLIENT_ATTRIBUTE);
```

Comment:
The `ServletContextHandler` here could be used before it's added to a `Server` handler tree, the above edit prevents the NPE.

Reply:
But if it used prior to being added to a ServerHandler tree, then will the HTTPCLIENT_ATTRIBUTE be checked again when it is?

Reply:
No.  This issue is about eagerness.

In normal standalone Jetty, this NPE isn't possible, as the normal SCI behaviors will kick in.  Making the check for `HTTPCLIENT_ATTRIBUTE` only occur once, on WebApp init.

In embedded-jetty, if they use this `WebSocketServerContainerInitializer.configureContext()` to get a JSR356 `ServerContainer` they are fixing in stone their usage, they want to start managing the `ServerContainer` immediately.  This could be as simple as just adding an endpoint, or configuring the client container within it.  There's also nothing stopping them from attempting to _use_ this `ServerContainer` to make a connection out to a remote WebSocket server too.

In embedded-jetty, they could alternatively use a `ServletContainerListener` to perform the init of specific webapps and the `ServerContainer` that's in the context attributes from within the webapp.
Or they could create an instance of `javax.websocket.server.ServerApplicationConfig` and have the startup of JSR356 use that to init their endpoints.

It's a conscious choice by the developer to use the optional `WebSocketServerContainerInitializer.configureContext()` when they did.   Repercussions abound.

As for the HttpClient, if the developer is eager, like the issue shows, then [the "default" HttpClient is used](https://github.com/eclipse/jetty.project/blob/jetty-9.4.14.v20181114/jetty-websocket/websocket-client/src/main/java/org/eclipse/jetty/websocket/client/HttpClientProvider.java).
Which means:

  1. If `XmlConfiguration` is available, then use `jetty-websocket-httpclient.xml` to load the `HttpClient` configuration
  2. Establish a new `HttpClient` for this specific `ServerContainer` -> `ClientContainer` -> `WebSocketClient` relationship.

An embedded-jetty user could also not care a lick about any of what I wrote above and just use ...

``` java
        org.eclipse.jetty.websocket.jsr356.server.ServerContainer container = WebSocketServerContainerInitializer.configureContext(context);
        container.addEndpoint(EchoSocket.class);
        container.addEndpoint(InfoSocket.class);

        HttpClient httpClient = container.getBean(HttpClient.class);
        httpClient.setIdleTimeout(300);
        httpClient.setMaxConnectionsPerDestination(5);
        httpClient.setMaxRedirects(3);
        httpClient.getProxyConfiguration().getProxies().add(new Socks4Proxy("proxy.corp.com", 7777));
```